### PR TITLE
Free used buffers in SHM when endpoint crashes [10992]

### PR DESCRIPTION
**Pull request by IkerLuengo**

### CODE REVIEW

File: src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp [hunk 1]
```diff
@@ -689,6 +692,46 @@ class SharedMemGlobal
             }
         }
 
+        bool get_and_remove_blocked_processing(
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp:695]: Document this new method

File: src/cpp/rtps/transport/shared_mem/SharedMemManager.hpp [hunk 2]
```diff
@@ -855,6 +878,26 @@ class SharedMemManager :
             return ret;
         }
 
+        bool recover_blocked_processing()
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/shared_mem/SharedMemManager.hpp:881]: The only place where this is used is ignoring the returned value. Consider returning void

File: src/cpp/rtps/transport/shared_mem/SharedMemManager.hpp [hunk 3]
```diff
@@ -855,6 +878,26 @@ class SharedMemManager :
...
+        {
+            bool recovered = false;
+            SharedMemGlobal::BufferDescriptor buffer_descriptor;
+            if (SharedMemGlobal::Port::is_zombie(global_port_->port_id(),
+                    shared_mem_manager_->global_segment()->domain_name()))
+            {
+                while (global_port_->get_and_remove_blocked_processing(buffer_descriptor))
+                {
+                    auto segment = shared_mem_manager_->find_segment(buffer_descriptor.source_segment_id);
+                    auto buffer_node =
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/shared_mem/SharedMemManager.hpp:891]: What happens if the segment is not found?

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp [hunk 4]
```diff
@@ -689,6 +707,63 @@ class SharedMemGlobal
...
+         *  - Finishes iterating
+         * 
+         * The intention of this method is to locate all buffer descriptors that are being processed
+         * and apply necessary operations on these buffers when the port is found to be zombie.
+         * This method should be called itearatively until the return value is false
+         * (there is no processing listener remaining).
+         * The calling method has the chance to apply necessary actions on the buffer descriptor.
+         * 
+         * @param[OUT] buffer_descriptor Descriptor of the buffer that was being processed by the first processing listener
+         * @return True if there was at least one processing listener. False if not.
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp:725]:
```suggestion
         * @brief Look for a listener processing a buffer and return the buffer descriptor being processed
         * 
         * Iterates over all the listeners, and upon finding the first one that is marked as \c is_processing:
         *  - Marks it as not processing
         *  - Copies the buffer descriptor that was being processed in the provided buffer
         *  - Finishes iterating
         * 
         * The intention of this method is to locate all buffer descriptors that are being processed
         * and apply necessary operations on these buffers when the port is found to be zombie.
         * This method should be called iteratively until the return value is false
         * (there is no processing listener remaining).
         * The calling method has the chance to apply necessary actions on the buffer descriptor.
         * 
         * @param[OUT] buffer_descriptor Descriptor of the buffer that was being processed by the first processing listener
         * @return True if there was at least one processing listener. False if not.
```

File: src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp [hunk 5]
```diff
@@ -689,6 +707,63 @@ class SharedMemGlobal
...
+        /**
+         * @brief Look for a listener processing a buffer and return the buffer descriptor being processing
+         * 
+         * Iterates over all the listeners, and upon founding the first one that is marked as \c is_processing:
+         *  - Marks it as not processing
+         *  - Copies the buffer descriptor that was being processed in the provided buffer
+         *  - Finishes iterating
+         * 
+         * The intention of this method is to locate all buffer descriptors that are being processed
+         * and apply necessary operations on these buffers when the port is found to be zombie.
```
Comment (MiguelCompany) [inline src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp:719]: Does this mean that a descriptor may be returned and the listener be marked as not processing even if the port is not zombie? If so, I think this is not the correct behavior.

Looking at this carefully, it seems that a precondition of this method is that the port is zombie. Please document that.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp (hunk 5 above)
Comment (IkerLuengo) [reply src/cpp/rtps/transport/shared_mem/SharedMemGlobal.hpp:719]: yes, that is a precondition. Documented.

File: test/communication/CMakeLists.txt [hunk 6]
```diff
@@ -110,8 +112,10 @@ if(NOT ((MSVC OR MSVC_IDE) AND EPROSIMA_INSTALLER) AND fastcdr_FOUND)
         ${CMAKE_CURRENT_BINARY_DIR}/secure_msg_submsg_crypto_besteffort_sub.xml COPYONLY)
     configure_file(${CMAKE_CURRENT_SOURCE_DIR}/secure_submsg_crypto_besteffort_pub.xml
         ${CMAKE_CURRENT_BINARY_DIR}/secure_submsg_crypto_besteffort_pub.xml COPYONLY)
-    configure_file(${CMAKE_CURRENT_SOURCE_DIR}/secure_submsg_crypto_besteffort_sub.xml
+        configure_file(${CMAKE_CURRENT_SOURCE_DIR}/secure_submsg_crypto_besteffort_sub.xml
```
Comment (MiguelCompany) [inline test/communication/CMakeLists.txt:115]:
```suggestion
    configure_file(${CMAKE_CURRENT_SOURCE_DIR}/secure_submsg_crypto_besteffort_sub.xml
```

File: test/communication/CMakeLists.txt [hunk 7]
```diff
@@ -362,10 +366,8 @@ if(NOT ((MSVC OR MSVC_IDE) AND EPROSIMA_INSTALLER) AND fastcdr_FOUND)
                 "PATH=$<TARGET_FILE_DIR:${PROJECT_NAME}>\\;$<TARGET_FILE_DIR:fastcdr>\\;${WIN_PATH}")
         endif()
 
-
-
         add_test(NAME SimpleCommunicationBestEffortFixed
-            COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/simple_communication.py)
+        COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/simple_communication.py)
```
Comment (MiguelCompany) [inline test/communication/CMakeLists.txt:370]:
```suggestion
            COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/simple_communication.py)
```

File: test/communication/shm_communication_subscriber_dies_while_processing_message.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+# Copyright 2016 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelCompany) [inline test/communication/shm_communication_subscriber_dies_while_processing_message.py:1]:
```suggestion
# Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Pull request merged**